allensdk/brain_observatory/behavior/session_apis/data_io/behavior_ophys_lims_api.py

```diff
@@ -0,0 +1,164 @@
+    LIMS_DB_CREDENTIAL_MAP, MTRAIN_DB_CREDENTIAL_MAP)
+from allensdk.core.authentication import credential_injector, DbCredentials
+from allensdk.brain_observatory.behavior.session_apis.data_transforms import (
+    BehaviorOphysDataXforms)
+
+
+class BehaviorOphysLimsApi(BehaviorOphysDataXforms,  OphysLimsApi,
+                           BehaviorLimsApi):
+
+    def __init__(self, ophys_experiment_id: int,
+                 lims_credentials: Optional[DbCredentials] = None,
+                 mtrain_credentials: Optional[DbCredentials] = None):
+
+        self.logger = logging.getLogger(self.__class__.__name__)
+
+        self.lims_db = db_connection_creator(
+            credentials=lims_credentials,
+            default_credentials=LIMS_DB_CREDENTIAL_MAP)
+
+        self.mtrain_db = db_connection_creator(
+            credentials=mtrain_credentials,
+            default_credentials=MTRAIN_DB_CREDENTIAL_MAP)
+
+        self.ophys_experiment_id = ophys_experiment_id
+        self.behavior_session_id = self.get_behavior_session_id()
+
+    def get_ophys_experiment_id(self):
+        return self.ophys_experiment_id
+
+    @memoize
+    def get_ophys_session_id(self):
+        query = '''
+                SELECT os.id FROM ophys_sessions os
+                JOIN ophys_experiment oe ON oe.ophys_session_id = os.id
+                WHERE oe.id = {};
+                '''.format(self.get_ophys_experiment_id())
+        return self.lims_db.fetchone(query, strict=True)
+
+    @memoize
+    def get_experiment_container_id(self):
+        query = '''
+                SELECT visual_behavior_experiment_container_id
+                FROM ophys_experiments_visual_behavior_experiment_containers
+                WHERE ophys_experiment_id= {};
+                '''.format(self.get_ophys_experiment_id())
+        return self.lims_db.fetchone(query, strict=False)
+
+    @memoize
+    def get_behavior_stimulus_file(self):
+        query = '''
+                SELECT stim.storage_directory || stim.filename AS stim_file
+                FROM ophys_experiments oe
+                JOIN ophys_sessions os ON oe.ophys_session_id = os.id
+                JOIN behavior_sessions bs ON bs.ophys_session_id=os.id
+                LEFT JOIN well_known_files stim ON stim.attachable_id=bs.id AND stim.attachable_type = 'BehaviorSession' AND stim.well_known_file_type_id IN (SELECT id FROM well_known_file_types WHERE name = 'StimulusPickle')
+                WHERE oe.id= {};
+                '''.format(self.get_ophys_experiment_id())
+        return safe_system_path(self.lims_db.fetchone(query, strict=True))
+
+    @memoize
+    def get_nwb_filepath(self):
+
+        query = '''
+                SELECT wkf.storage_directory || wkf.filename AS nwb_file
+                FROM ophys_experiments oe
+                LEFT JOIN well_known_files wkf ON wkf.attachable_id=oe.id AND wkf.well_known_file_type_id IN (SELECT id FROM well_known_file_types WHERE name = 'BehaviorOphysNwb')
```

Comment:
carriage returns for readable

Reply:
also, btw, i don't think you need nested SELECT here. I think you could do this as:
```
SELECT wkf.storage_directory || wkf.filename AS nwb_file
FROM ophys_experiments oe
JOIN well_known_files wkf ON wkf.attachable_id=oe.id
JOIN well_known_file_types wkft ON wkf.well_known_file_type_id=wkft.id
WHERE wkft.name = 'BehaviorOphysNwb'
```

Reply:
Many of these queries have now been reworked to take on the form suggested!